sashimi/gui/camera_gui.py

```diff
@@ -47,237 +61,260 @@ class ViewingWidget(QWidget):
         )
 
     def refresh(self) -> None:
         """Main refresh loop called by timeout of the main timer."""
         self.refresh_image()
-        self.refresh_progress_bar()
 
     def refresh_image(self):
         current_image = self.state.get_volume()
         if current_image is None:
             return
 
-        delta_t = (time_ns() - self.last_time_updated) / 1e9
-
-        if delta_t > 1 / self.display_settings.display_framerate:
-            # If not volumetric or out of range, reset indexes:
-            if current_image.shape[0] == 1:
-                self.frame_layer.dims.reset()
-            self.frame_layer.data = current_image
-            self.frame_layer.scale = [self.voxel_size[0] / self.voxel_size[1], 1.0, 1.0]
-            self.last_time_updated = time_ns()
-
-    def refresh_progress_bar(self):
-        sstatus = self.state.get_save_status()
-        if sstatus is not None:
-            self.experiment_progress.show()
-            self.lbl_experiment_progress.show()
-            self.experiment_progress.setMaximum(sstatus.target_params.n_volumes)
-            self.experiment_progress.setValue(sstatus.i_volume)
-            self.lbl_experiment_progress.setText(
-                "Saved chunks: {}".format(sstatus.i_chunk)
-            )
-
-    def update_contrast_limits(self):
-        self.frame_layer.reset_contrast_limits()
-
-    def toggle_roi_display(self):
-        if self.roi.visible:
-            # TODO: Block rotating roi
-            self.roi.visible = False
+        # If not volumetric or out of range, reset indexes:
+        if current_image.shape[0] == 1:
+            self.frame_layer.dims.reset()
+        self.frame_layer.data = current_image
+        self.frame_layer.scale = [self.voxel_size[0] / self.voxel_size[1], 1.0, 1.0]
+
+        # Check if anything changed in the image shape, which would mean that changes of the contrast
+        # are required (in case the parameter update was missed).
+        if self.is_first_frame or self.image_shape != current_image.shape:
+            self.reset_contrast()
+            self.viewer.reset_view()
+
+        self.image_shape = current_image.shape
+        self.is_first_frame = False
+
+    def reset_contrast(self):
+        if self.auto_contrast:
+            self.frame_layer.reset_contrast_limits()
         else:
-            self.roi.visible = True
-            self.viewer.press_key(
-                "s"
-            )  # this allows moving the whole roi and scaling but no individual handles
+            self.set_manual_contrast()
 
-    def update_roi(self):
-        pass
+    def auto_contrast(self):
+        if self.auto_contrast_chk.isChecked():
+            self.wid_contrast_range.hide()
+            self.auto_contrast = True
+        else:
+            self.wid_contrast_range.show()
+            self.auto_contrast = False
+
+    def set_manual_contrast(self):
+        self.frame_layer.contrast_limits = [i for i in self.contrast_range.contrast_range]
```

Comment:
why the list comprehension? what is .contrast_range so that it has to be converted in this involved way?